### DISCUSSION TITLE
making items unique in HTML+XPath

### DISCUSSION THREAD

**Discussion by rezad1393** [FreshRSS/FreshRSS#5824] [Q&A] [answered]
I have this feed with type HTML+XPath:

```
      <outline text="Youth in Cinema" type="HTML+XPath" xmlUrl="https://www.youthincinema.com/index.php" htmlUrl="https://www.youthincinema.com/index.php" description="RSS feed of Youth in Cinema" frss:xPathItem="//div[@class=&quot;movieinfo&quot;]" frss:xPathItemTitle="descendant::b[1]/a" frss:xPathItemContent="." frss:xPathItemUri="descendant::a/@href" frss:xPathItemTimestamp="descendant::b[contains(text(),'Movie added at')]/following-sibling::text()[1]" frss:xPathItemThumbnail="descendant::img/@src"/>
    </outline>
```

I fine tuned it to have date extracted from post and so on.

but every time a new items gets added all of max number of post on the page get shown as new post.

how I can make freshrss not show past posts as new ones?
is it related to 
```
XPath (relative to item) for:
item unique ID
```

I remember that when I was trying to fine tune the feed and I would put a wrong one for titles it would show something like 
URN:hash-a-lot-of-characters

is that a hash?
of the post?

can I use that for unique ids?

does anyone here has experience with this xpath feature of freshrss that can guide me?

**Comment by Alkarex** [comment, marked as answer]
Quick answer already: make sure to fill the *item unique ID* field. You can use the same expression as for your *item link (URL)*. This will help existing posts not being added again.

**Reply by rezad1393** [reply]
I am too dumb.

should have though of that.

thanks.


is there a place for examples of this type of feeds?

I am not a web dev but I can copy/paste and grok a bit.



p.s. : forgot discussions exist and made a bug report. my bad.

**Reply by rezad1393** [reply]
btw do you know what is that?:
urn:sha1:037b85ab4148e136eed420204daefe636e7454ad

**Reply by Alkarex** [reply]
See related XPath questions on https://github.com/FreshRSS/FreshRSS/discussions?discussions_q=xpath , some of which have use-cases / examples.

But the *unique ID* field was added a bit later https://github.com/FreshRSS/FreshRSS/pull/4507 so not all examples take advantage of it

`urn:sha1:037b85ab4148e136eed420204daefe636e7454ad` is a unique identifier, or guid, using a hash of the content.
See:
* https://en.wikipedia.org/wiki/Uniform_Resource_Name
* https://en.wikipedia.org/wiki/SHA-1
* https://www.rssboard.org/rss-specification#ltguidgtSubelementOfLtitemgt

**Reply by rezad1393** [reply]
`urn:sha1:037b85ab4148e136eed420204daefe636e7454ad is a unique identifier, or guid, using a hash of the content.`

what content exactly?

when I put a bad option for title like "frss:xPathItemTitle="descendant::**ssssb**[1]/a" instead of frss:xPathItemTitle="descendant::**b**[1]/a

I get that hashes for title.

what exactly happens that title gets that hash as the answer?

**Reply by Alkarex** [reply]
An article must have an ID and a title. If not provided correctly, FreshRSS will generate an ID automatically based on the content. That ID will also be used as the title if the title is empty.

Here is the relevant part of the code:
https://github.com/FreshRSS/FreshRSS/blob/711e2153d11f3df181be2f6248368d75a36b5725/app/Models/Feed.php#L718-L720

**Reply by rezad1393** [reply]
is that ID the same as the unique ID that I had to add to end of option of feed?

if so then why I got duplicates.
if not can I use that guid directly for feed options ? like instead of using link for unique id can I put guid that is computed? what parameters should I use for that?

I am aware that guid changes if those title.content.link change, but I am curious to use it.

and thank you for the replies.
can you put these somewhere in wiki so others can you them too?

**Reply by Alkarex** [reply]
If you do not provide the ID yourself, a unique hash is computed. The duplicates are probably due to a change in the content.

**Reply by Alkarex** [reply]
PR to our documentation welcome ;-)
https://github.com/FreshRSS/FreshRSS/blob/edge/docs/en/users/11_website_scraping.md

**Comment by rezad1393** [comment]
> If you do not provide the ID yourself, a unique hash is computed. The duplicates are probably due to a change in the content.
are you sure?

that website doesn't usually change last post when new posts are done.

but maybe it is because  some of the "." content is html stuff and changes when posting new divs. 

though I tested with starring some items of that feed and didnt see stuff like ".panel:nth-of-type(4)" in the posts, that other sites use to arrange posts and that would change when posting new stuff.

right now using url for unique id is good and I will see if it works when new post are added to the site and thank you for pointing it out to me ( I can't believe I didn't see that one :) ) but I was really curios  abou that hash thing.

 I think my whole life I have done two PR merge request and that was with them taking my hand and helping me step by step. but what long ago.
also I usually write my stuff informally and may not be appropriate for wiki but I may try and you be the judge.


btw I wish freshrss would separate docs in  another repo for better management so that I dont have to fork this whole repo.( that's the way to PR right? first fork then edit?)

thank you and I think I close this.

**Reply by Alkarex** [reply]
Example of PR process in a few clicks from your Web browser:
* Go to https://github.com/FreshRSS/FreshRSS/blob/edge/docs/en/users/11_website_scraping.md
* Click on the *edit* icon and follow the instructions (yes, it automatically makes a fork)
![image](https://github.com/FreshRSS/FreshRSS/assets/1008324/efc69eea-ee86-4d62-b3d8-f79b7362c73b)
* Type some text and use the *preview* to see how it looks like
* When it looks fine, follow the instructions and you will end up with a functioning PR
* We (and you) can then continue doing additional edits until the PR is good enough (without having to make a new PR)
